### PR TITLE
Add containerd windows support on GCE for test.

### PR DESCRIPTION
**Pull request by Random-Liu**
Ref https://github.com/kubernetes/enhancements/issues/1001
Ref https://github.com/containerd/cri/issues/1257

Add containerd windows support on GCE for testing.

No doc or release note change yet because this is only for testing. We can make this official after a release is cut in containerd.

This PR:
1) Installed crictl on windows for both Docker and containerd;
2) Installed containerd from the test bucket `gs://cri-containerd-staging/windows` for testing;
3) Updated pause image version to 1.2.0 for port forward support /cc @yliaog We may want to do this for 1.17, or else port forward will not work.
4) Installed multi-format-parser plugin for stackdriver and added CRI logging format support /cc @barney-s 
5) Containerd CNI uses `sdnbridge.exe` instead of `win-bridge.exe`. `win-bridge.exe` has very docker specific logic that doesn't work with containerd (see [here](https://github.com/containernetworking/plugins/blob/a11cb626b0c136d900eec642bb0e59f4cc365e73/pkg/hns/endpoint_windows.go#L30)). /cc @lzang 
6) Moved `hostdns.conf` from `CNI_CONFIG_DIR` to `CNI_DIR`. We shouldn't put arbitrary config into `CNI_CONFIG_DIR`, runtime may think it is a CNI config and return error (at least containerd will). /cc @lzang 

I'll add `containerd.log` fluentd and rotation support once https://github.com/containerd/containerd/pull/3831 is merged.

To try this change, follow the [README.md](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/windows/README-GCE-Windows-kube-up.md). The only change would be **`export KUBE_CONTAINER_RUNTIME=containerd` before kube-up**.

/cc @yujuhong @pjh @yliaog @kubernetes/sig-windows-misc  @kubernetes/sig-node-pr-reviews 

Signed-off-by: Lantao Liu <lantaol@google.com>

```release-note
none
```

### CODE REVIEW

**Review by pjh — COMMENTED** [review]
This looks great, thanks Lantao!

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 1]
```diff
@@ -51,7 +51,7 @@
 #  - Document functions using proper syntax:
 #    https://technet.microsoft.com/en-us/library/hh847834(v=wps.620).aspx
 
-$INFRA_CONTAINER = 'gcr.io/gke-release/pause-win:1.0.0'
+$INFRA_CONTAINER = 'mcr.microsoft.com/k8s/core/pause:1.2.0'
```
Comment (pjh) [inline cluster/gce/windows/k8s-node-setup.psm1]: @yliaog does `gke-release/pause-win:1.0.0` already include the port-forwarding support?

If so then is this change necessary?

File: cluster/gce/windows/common.psm1 [hunk 2]
```diff
@@ -106,35 +106,41 @@ function Get-InstanceMetadataAttribute {
...
 # successful, it will write the file to OutFile. You can optionally provide a
-# SHA1 Hash argument, in which case it will attempt to validate the downloaded
-# file against the hash.
+# Hash argument with an optional Algorithm, in which case it will attempt to
+# validate the downloaded file against the hash. SHA1 will be used if Algorithm
+# is not provided.
 function MustDownload-File {
   param (
     [parameter(Mandatory=$false)] [string]$Hash,
+    [parameter(Mandatory=$false)] [string]$Algorithm,
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply cluster/gce/windows/common.psm1]: Done

File: cluster/gce/windows/configure.ps1 [hunk 3]
```diff
@@ -131,15 +131,25 @@ try {
...
   Create-KubeletKubeconfig
   Create-KubeproxyKubeconfig
   Set-PodCidr
   Configure-HostNetworkingService
-  Configure-CniNetworking
+  if (${env:CONTAINER_RUNTIME} -eq "containerd") {
+    Configure-Containerd-CniNetworking
+  } else {
+    Configure-CniNetworking
+  }
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply cluster/gce/windows/configure.ps1]: Done

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 4]
```diff
@@ -1115,27 +1120,68 @@ function Verify-WorkerServices {
   Log_Todo "run more verification commands."
 }
 
+function DownloadAndInstall-Crictl {
+  $CRICTL_VERSION = "v1.16.1"
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: I think this meant to be hard coded.

It should match the crictl version in the deb, rpm package and also linux bootstrap script on GCE.

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 5]
```diff
@@ -1115,27 +1120,68 @@ function Verify-WorkerServices {
...
 # TODO(pjh): downloading the container container image may take a few minutes;
 # figure out how to run this in the background while perform the rest of the
 # node startup steps!
 function Pull-InfraContainer {
+  $COMMAND = "docker"
+  $LIST = "images"
+  $PULL = "pull"
+  $INSPECT = "inspect"
+  if (${env:CONTAINER_RUNTIME} -ne "docker") {
+    $COMMAND = "crictl"
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: Crictl talks with dockershim instead of dockerd. So once kubelet is started, crictl will work.

However, here kubelet hasn't started yet, we have to use docker.

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 6]
```diff
@@ -1115,27 +1120,68 @@ function Verify-WorkerServices {
...
+      'download/' + $CRICTL_VERSION + '/crictl-' +
+      $CRICTL_VERSION + '-windows-amd64.tar.gz')
+  MustDownload-File `
+      -URLs $url `
+      -OutFile $tmp_dir\crictl.tgz `
+      -Hash $CRICTL_SHA256 `
+      -Algorithm SHA256
+
+  Push-Location $tmp_dir
+  # tar can only extract in the current directory.
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: > TODO(pjh): clean this up, files not guaranteed to end up in NODE_DIR

I copied the CNI code, which isn't using `-C`. :)

> Relatedly, at line 350 could you add TODO(pjh): clean this up, files not guaranteed to end up in NODE_DIR ? That code is poorly written.

Done

**Comment by lzang** [comment]
Lantao, can we do the hostdns.conf path change in a separate pr? Thanks.

**Comment by yliaog** [comment]
> Lantao, can we do the hostdns.conf path change in a separate pr? Thanks.

+1, it's better to have it separately, and for easy cherrypicking

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 7]
```diff
@@ -839,6 +840,16 @@ Unblock-File $modulePath
 Import-Module -Name $modulePath'.replace('K8S_DIR', ${env:K8S_DIR})
 }
 
+# Setup cni network. This function supports both Docker
+# and containerd.
+function Configure-CniNetworking {
+  if (${env:CONTAINER_RUNTIME} -eq "containerd") {
+    Configure-Containerd-CniNetworking
```
Comment (pjh) [inline cluster/gce/windows/k8s-node-setup.psm1]: Internal functions should be named with underscores, which prevents them from being made visible when this module is imported.

`Configure_Containerd_CniNetworking` etc.

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 1 above)
Comment (pjh) [reply cluster/gce/windows/k8s-node-setup.psm1]: @yliaog do you know?

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (pjh) [reply cluster/gce/windows/k8s-node-setup.psm1]: Ah, I see. Can you add a comment explaining that?

**Comment by pjh** [comment]
LGTM with the change that lzang suggested.

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: Actually we moved PullInfraContainer to the end after kubelet started, so I think we should be able to use crictl now.

I'll still add comment to emphasize that the order needs to be kept in the future.

**Comment by Random-Liu** [comment]
> Lantao, can we do the hostdns.conf path change in a separate pr? Thanks.

I can move it to a separate commit, but we should make it a separate PR, or else this PR won't work at all. :)

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 1 above)
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: It seems not:
```
PS C:\> wincat
wincat : The term 'wincat' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ wincat
+ ~~~~~~
    + CategoryInfo          : ObjectNotFound: (wincat:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

I got above from a `gcr.io/gke-release/pause-win:1.0.0` container

Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: It is in the image, just not in PATH in powershell for some reason.

I'll remove the image update. :)

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 7 above)
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: Done

**Comment by Random-Liu** [comment]
> LGTM with the change that lzang suggested.

Moved that into a second commit.

**Comment by pjh** [comment]
/lgtm
/approve

pull-kubernetes-e2e-windows-gce has been flaky recently, if the cluster comes up successfully but a few tests fail then it should be fine. If the cluster fails to come up then please investigate the Windows node startup.

**Comment by pjh** [comment]
/approve

for `cluster/gce/windows/`

**Comment by yujuhong** [comment]
/approve

for the GCE directory

**Comment by BenTheElder** [comment]
/approve
for build

File: cluster/gce/windows/k8s-node-setup.psm1 [hunk 8]
```diff
@@ -1115,28 +1131,74 @@ function Verify-WorkerServices {
...
+
+  # Assume that presence of crictl.exe indicates that the crictl binaries
+  # were already previously downloaded to this node.
+  if (-not (ShouldWrite-File ${env:NODE_DIR}\crictl.exe)) {
+    return
+  }
+  $tmp_dir = 'C:\crictl_tmp'
+  New-Item $tmp_dir -ItemType 'directory' -Force | Out-Null
+
+  $url = ('https://github.com/kubernetes-sigs/cri-tools/releases/' +
```
Comment (yliaog) [inline cluster/gce/windows/k8s-node-setup.psm1]: could you put it in a gs bucket? similarly to other binaries, like stackdriver, or cni

File: cluster/gce/util.sh [hunk 9]
```diff
@@ -878,7 +878,7 @@ function construct-windows-kubelet-flags {
   flags+=" --logtostderr=false"
 
   # Configure the file path for host dns configuration
-  flags+=" --resolv-conf=${WINDOWS_CNI_CONFIG_DIR}\hostdns.conf"
+  flags+=" --resolv-conf=${WINDOWS_CNI_DIR}\hostdns.conf"
```
Comment (yliaog) [inline cluster/gce/util.sh]: this commit is still here, could you move it to a separate pr?

Comment (Random-Liu) [reply cluster/gce/util.sh]: See the comment https://github.com/kubernetes/kubernetes/pull/85396#issuecomment-559217726

File: cluster/gce/windows/k8s-node-setup.psm1 (hunk 8 above)
Comment (Random-Liu) [reply cluster/gce/windows/k8s-node-setup.psm1]: CNI is not in a gcs bucket.

Comment (yliaog) [reply cluster/gce/windows/k8s-node-setup.psm1]: if it is not now, it needs to be released in gcs soon.

Comment (BenTheElder) [reply cluster/gce/windows/k8s-node-setup.psm1]: Why?

Comment (riking) [reply cluster/gce/windows/k8s-node-setup.psm1]: GitHub does have download quotas; using GCS ensures usage is billed instead of blocked

Comment (BenTheElder) [reply cluster/gce/windows/k8s-node-setup.psm1]: production users should probably mirror things they will pull very frequently sure, but github is 100% reasonable for upstream binary releases.
